shtdlib.sh

```diff
@@ -141,6 +141,23 @@ elif [ "${os_family}" == 'Alpine' ]; then
     os_name='alpine'
 fi
 
+# Encode the given string
+# We encode by substituting the hex unicode of special characters
+function url_encode {
+    local url_length=${#1}
+    for ((i=0 ; i<url_length ; i++)); do
+        char=${1:i:1}
+        case "${char}" in
+            [a-zA-Z0-9.~_-])
+                printf "%s" "${char}" ;;
+            ' ')
+                printf + ;;
+            *)
+                printf %%%X "'${char}" ;;
```

Comment:
OK a couple questions about the case part of this: 

1. what about other special characters... like +, %, ^ etc?
2. Why are we replacing whitespace with `+`? I thought the url encoding of + was %20?
3. Similar to 2, i'm confused as to why we are replacing * with %%%X*? I figured it'd be replaced with 2A. 
What am i misunderstanding here about the way we want to encode this?

Reply:
1. those will be uncoded via `printf %%%X "'${char}" ;`
2. here is a good answer to 2nd question https://stackoverflow.com/questions/1634271/url-encoding-the-space-character-or-20. basically + is safer to encode space char
3. so formatting a string in printf you do something similar to what we do in C. basically ("%s", "hello") it will format the %s with hello. in our case we want the `%` char to be there, so we escape the first percent with %, hence %%. then we need hex value of the unicoded character, which is %X, hence we get %%%X.

Reply:
1. OH so the '*' on the last case is actually a regular expression, catching any other character, correct?
2. Thanks that helps! 
3. Still thinking about this... I think if i give it a few minutes it'll start to make sense. Will report back.

Reply:
Kevin explained over slack that when using prinf, putting a single quote before a character will give you the unicode representation of it.